url41e0dd6de9_jakaldama_pipeline_simulator_tgz-pJ8-PipelineSimulatorJ8/src/PipelineSimulator.java

```diff
@@ -0,0 +1,590 @@
+			}
+			else if (IDEX.getReadALUSrc() == 0)
+				EXMEM.setWriteALUresult(IDEX.getReadReadData1() + IDEX.getReadReadData2());
+		} else if (IDEX.getReadALUcontrol() == 6) {
+			//this must be a subtract operation
+			EXMEM.setWriteALUresult(IDEX.getReadReadData1() - IDEX.getReadReadData2());
+		}
+
+		//Use RegDst to set which register will be written to with ALU executed product
+		if (IDEX.getReadRegDst() == 0)
+			EXMEM.setWriteWriteRegister(EXMEM.getWriteRegT());
+		else
+			EXMEM.setWriteWriteRegister(EXMEM.getWriteRegD());
+	}
+	
+	public static void MEM_stage() {
+		//Needed for WB stage
+		MEMWB.setWriteWriteRegister(EXMEM.getReadWriteRegister());
+		MEMWB.setWriteALUresult(EXMEM.getReadALUresult());
+		//send regWrite and MemtoReg to MEMWB register
+		MEMWB.setWriteRegWrite(EXMEM.getReadRegWrite());
+		MEMWB.setWriteMemtoReg(EXMEM.getReadMemtoReg());
+		//instruction data
+		MEMWB.setWriteInstr(EXMEM.getReadInstr());
+		MEMWB.setWriteRegD(EXMEM.getReadRegD());
+		MEMWB.setWriteRegS(EXMEM.getReadRegS());
+		MEMWB.setWriteRegT(EXMEM.getReadRegT());
+		MEMWB.setWriteOffset(EXMEM.getReadOffset());
+		
+		//Check control signal MemRead, if asserted, this is a load - load data, prepare for write back stage
+		if (EXMEM.getReadMemRead() == 1) {
+			MEMWB.setWriteData(mainMemory[(MEMWB.getWriteALUresult() & 0xFF)]);
+			
+		}
+		//Check control signal MemWrite, if asserted, this is a store - store data into mainMemory now
+		if (EXMEM.getReadMemWrite() == 1) {
+			mainMemory[MEMWB.getWriteALUresult()] = regs[MEMWB.getWriteRegT()];
+		}
+		
+
+	}
+	
+	public static void WB_stage() {
+		//RegWrite asserted if anything but store, which instruction would already be complete by now anyway
+		//Check MemToReg, if not asserted, value fed to Write register comes from the ALU product
+		//If MemtoReg is asserted, value fed to Write register comes from data memory
+		if (MEMWB.getReadRegWrite() == 1) {
+			if (MEMWB.getReadMemtoReg() == 0) {
+				//Add or subtract
+				regs[MEMWB.getReadWriteRegister()] = MEMWB.getReadALUresult();
+			//value fed to register Write data input comes from the ALU
+			} else if (MEMWB.getReadMemtoReg() == 1) {
+				//This is a load- value fed to register Write data input comes from the data memory, stored as Write/Read Data in MEMWB reg
+				regs[MEMWB.getReadWriteRegister()] = MEMWB.getReadData();
+			}
+		}
+	}
+	
+	public static String opName(int i) {
+		int tempOpcode = (i >>> 26) & 0x3F;
+		String opName = "";
+		switch (tempOpcode) {
+		case 32 :
+			opName = "lb";
+			break;
+		case 40:
+			opName = "sb";
+			break;
+			
+		}
+		return opName;
+			
+	}
+	
+	public static String functionName(int i) {
+		int tempFunction = i & 0x3F;
+		String funcName = "";
+		switch (tempFunction) {
+		case 32:
+			funcName = "add";
+			break;
+		case 34:
+			funcName = "sub";
+			break;
+		}
+		return funcName;
+	}
+
+	public static void Print_out_everything() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 112 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks such as on lines 153-163, lines 166-176, lines 182-198, lines 219-237, lines 258-276 into separate methods.